Modules/_sqlite/statement.c

```diff
@@ -48,13 +48,43 @@ typedef enum {
     TYPE_UNKNOWN
 } parameter_type;
 
+static int pysqlite_statement_is_dml(sqlite3_stmt *statement, const char *sql)
+{
+    const char* p;
+    int is_dml = 0;
+
+    is_dml = !sqlite3_stmt_readonly(statement);
+    if (is_dml) {
+        /* Retain backwards-compatibility, as sqlite3_stmt_readonly will return
+         * false for BEGIN [IMMEDIATE|EXCLUSIVE] or DDL statements. */
+        for (p = sql; *p != 0; p++) {
+            switch (*p) {
+                case ' ':
+                case '\r':
+                case '\n':
+                case '\t':
+                    continue;
+            }
+
+            is_dml = (PyOS_strnicmp(p, "begin", 5) &&
+                      PyOS_strnicmp(p, "create", 6) &&
+                      PyOS_strnicmp(p, "drop", 4) &&
+                      PyOS_strnicmp(p, "alter", 5) &&
+                      PyOS_strnicmp(p, "analyze", 7) &&
+                      PyOS_strnicmp(p, "reindex", 7) &&
+                      PyOS_strnicmp(p, "vacuum", 6));
```

Comment:
I don't think we need all of those.

- `sqlite3_stmt_readonly()` with `ALTER` statement returns `0`
- `sqlite3_stmt_readonly()` with `VACUUM` statement returns `0`
- `sqlite3_stmt_readonly()` with `ANALYZE` statement returns `0`
- `sqlite3_stmt_readonly()` with `DROP` statement returns `0`
- `sqlite3_stmt_readonly()` with `CREATE` statement that actually creates a new item returns `0`
- `sqlite3_stmt_readonly()` with `CREATE` statement that resolves to a no-op returns `1`